csharp/src/Ice/IceDiscovery/Locator.cs

```diff
@@ -12,229 +12,470 @@
+                            {
+                                _replyAdapter.Communicator.Logger.Warning(ex.ToString());
+                                replyHandler.SetEmptyResult();
+                            }
+                        }
+                    }
+
+                    Task? t = await Task.WhenAny(replyHandler.Task,
+                        Task.Delay(_timeout, replyHandler.CancellationToken)).ConfigureAwait(false);
+                    if (t == replyHandler.Task)
+                    {
+                        return await replyHandler.Task.ConfigureAwait(false);
+                    }
+                    else if (t.IsCanceled)
                     {
-                        adapterIds = new HashSet<string>();
-                        _replicaGroups.Add((replicaGroupId, protocol), adapterIds);
+                        // If the timeout was canceled we delay the completion of the request to give a chance to other
+                        // members of this replica group to reply
+                        return await replyHandler.WaitForReplicaGroupRepliesAsync(start, _latencyMultiplier)
+                            .ConfigureAwait(false);
                     }
-                    adapterIds.Add(adapterId);
+                    // else timeout, so we retry until _retryCount
                 }
+                replyHandler.SetEmptyResult(); // Timeout
+                return await replyHandler.Task.ConfigureAwait(false);
+            }
+            finally
+            {
+                _replyAdapter.Remove(replyIdentity);
             }
         }
+    }
 
-        public void SetAdapterDirectProxy(
-            string adapterId,
-            IObjectPrx? proxy,
-            Current current,
-            CancellationToken cancel)
-        {
-            Debug.Assert(false); // this method is never called since this servant is hosted by an ice2 object adapter.
-        }
+    /// <summary>Helps collect / gather the reply(ies) to a lookup request.</summary>
+    internal sealed class ReplyHandler<TResult> : IDisposable
+    {
+        internal CancellationToken CancellationToken => _cancellationSource.Token;
+        internal Task<TResult> Task => _completionSource.Task;
+        internal IObject Servant { get; }
+
+        private readonly CancellationTokenSource _cancellationSource;
+        private readonly TaskCompletionSource<TResult> _completionSource;
 
-        public void SetReplicatedAdapterDirectProxy(
-           string adapterId,
-           string replicaGroupId,
-           IObjectPrx? proxy,
-           Current current,
-           CancellationToken cancel)
+        private readonly Func<TResult> _emptyResult;
+        private readonly Func<TResult>? _replicaReplyCollector;
+
+        public void Dispose() => _cancellationSource.Dispose();
+
+        internal ReplyHandler(IObject servant, Func<TResult> emptyResult, Func<TResult>? replicaReplyCollector = null)
```

Comment:
I think we can pass the default directly, don't need to use a function, the default is either a null proxy and empty list or an empty string

Reply:
I don't think that's true. There is no default for `string` or `IReadOnlyList<EndpointData>`.

Reply:
I mean you can use `null`, `""`, `IReadOnlyList<EndpointData>.Empt`y without having to create these functions

Reply:
Yes, you're right. Fixed.